extensions-core/namespace-lookup/src/main/java/io/druid/query/lookup/NamespaceLookupExtractorFactory.java

```diff
@@ -0,0 +1,292 @@
+import com.metamx.common.logger.Logger;
+import io.druid.query.extraction.MapLookupExtractor;
+import io.druid.query.lookup.namespace.ExtractionNamespace;
+import io.druid.server.lookup.namespace.cache.NamespaceExtractionCacheManager;
+
+import javax.annotation.Nullable;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+@JsonTypeName("cachedNamespace")
+public class NamespaceLookupExtractorFactory implements LookupExtractorFactory
+{
+  private static final Logger LOG = new Logger(NamespaceLookupExtractorFactory.class);
+
+  private static final long DEFAULT_SCHEDULE_TIMEOUT = 60_000;
+  private static final byte[] CLASS_CACHE_KEY;
+
+  static {
+    try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
+      baos.write(StringUtils.toUtf8(NamespaceLookupExtractorFactory.class.getCanonicalName()));
+      baos.write(0xFF);
+      CLASS_CACHE_KEY = baos.toByteArray();
+    }
+    catch (IOException e) {
+      // Should never happen
+      throw Throwables.propagate(e);
+    }
+  }
+
+  @JsonIgnore
+  private final AtomicBoolean started = new AtomicBoolean(false);
+  @JsonIgnore
+  private final ReadWriteLock startStopSync = new ReentrantReadWriteLock();
+  @JsonIgnore
+  private final NamespaceExtractionCacheManager manager;
+  @JsonIgnore
+  private final LookupIntrospectHandler lookupIntrospectHandler;
+  private final ExtractionNamespace extractionNamespace;
+  private final long firstCacheTimeout;
+  private final boolean oneToOne;
+
+  private final String extractorID;
+
+  @JsonCreator
+  public NamespaceLookupExtractorFactory(
+      @JsonProperty("extractionNamespace") ExtractionNamespace extractionNamespace,
+      @JsonProperty("firstCacheTimeout") Long firstCacheTimeout,
+      @JsonProperty("oneToOne") boolean oneToOne,
+      @JacksonInject final NamespaceExtractionCacheManager manager
+  )
+  {
+    this.extractionNamespace = Preconditions.checkNotNull(
+        extractionNamespace,
+        "extractionNamespace should be specified"
+    );
+    this.firstCacheTimeout = firstCacheTimeout == null ? DEFAULT_SCHEDULE_TIMEOUT : firstCacheTimeout;
+    Preconditions.checkArgument(this.firstCacheTimeout >= 0);
+    this.oneToOne = oneToOne;
+    this.manager = manager;
+    this.extractorID = buildID();
+    this.lookupIntrospectHandler = new LookupIntrospectHandler()
+    {
+      @GET
+      @Path("/keys")
+      @Produces(MediaType.APPLICATION_JSON)
+      public Response getKeys()
+      {
+        return Response.ok(getLatest().keySet().toString()).build();
```

Comment:
this is returning always ok response even if weird thing happens at `getLastest`, i think you need to catch the cases where that method fails